src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs

```diff
@@ -0,0 +1,350 @@
+                    newMethodBody = newMethodBody.Update(newMethodBody.Locals, newMethodBody.LocalFunctions, newStatements);
+
+                    if (IsTestMethod(method))
+                    {
+                        // If the method is a test method, wrap the body in:
+                        //
+                        // Instrumentation.FlushPayload();
+                        // try
+                        // {
+                        //     ... body ...
+                        // }
+                        // finally
+                        // {
+                        //     Instrumentation.FlushPayload();
+                        // }
+
+                        BoundStatement firstFlush = factory.ExpressionStatement(factory.Call(null, flushPayload));
+                        BoundStatement secondFlush = factory.ExpressionStatement(factory.Call(null, flushPayload));
+                        BoundStatement tryFinally = factory.Try(newMethodBody, ImmutableArray<BoundCatchBlock>.Empty, factory.Block(ImmutableArray.Create(secondFlush)));
+                        newMethodBody = factory.Block(ImmutableArray.Create(firstFlush, tryFinally));
+                    }
+
+                    return newMethodBody;
+                }
+            }
+
+            dynamicAnalysisSpans = ImmutableArray<SourceSpan>.Empty;
+            return methodBody;
+        }
+
+        private static FieldSymbol GetPayloadField(MethodSymbol method, int methodOrdinal, TypeSymbol payloadType, SyntheticBoundNodeFactory factory, CSharpCompilation compilation)
+        {
+            // ToDo: If the type containing the method is generic, synthesize a helper type and put the payload field there.
+            // If the payload field is part of a generic type, there will be a new instance of the field per instantiation of the generic,
+            // and so the payload field must be a member of another type.
+            NamedTypeSymbol containingType = method.ContainingType;
+
+            SynthesizedFieldSymbol payloadField = new SynthesizedFieldSymbol(containingType, payloadType, method.Name + "*instrumentation*" + methodOrdinal.ToString(), isStatic: true);
+            factory.AddField(containingType, payloadField);
+            return payloadField;
+        }
+        
+        private static MethodSymbol GetCreatePayload(CSharpCompilation compilation)
+        {
+            if (_createPayload == null)
+            {
+                NamedTypeSymbol instrumentationType = compilation.GetTypeByMetadataName("Microsoft.CodeAnalysis.Runtime.Instrumentation");
+                if (instrumentationType != null)
+                {
+                    ImmutableArray<Symbol> createPayloads = instrumentationType.GetMembers("CreatePayload");
+                    if (createPayloads.Length == 1)
+                    {
+                        MethodSymbol createPayload = createPayloads[0] as MethodSymbol;
+                        // ToDo: Add checks for parameter types.
```

Comment:
Please include a bug number in each TODO.  In this case, the blanked bug for implementing instrumentation would suffice.